File: Algo/Leetcode/496NextGreaterNum.cpp
// dict + 单调栈
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        int n = nums1.size();
        int m = nums2.size();
        map<int, int> d;
        for(int i:nums2) d[i] = -1;

        stack<int> s;
        for(int i=0; i<m; i++){
             while(!s.empty() && nums2[s.top()] < nums2[i]){
                 d[nums2[s.top()]] = nums2[i];
                 s.pop();
             }
             s.push(i);
        }

        vector<int> res;
        for(int i:nums1){
            if(d[i]==-1) res.push_back(-1);
            else res.push_back(d[i]);
        }
        return res;
    }
};

File: Algo/Leetcode/496NextGreaterNum.py
class Solution(object):
    def nextGreaterElement(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """

        d = {}
        for n in nums2:
            d[n] = -1

        st = []
        for n in nums2:
            while len(st) > 0 and st[-1] < n:       # all the top on stack meet next larger n
                t = st.pop()
                d[t] = n
            st.append(n)

        res = []
        for n in nums1:
            res.append(d[n])

        return res